src/util.js

```diff
@@ -276,6 +281,86 @@ function copyDescriptors(from, to, filter) {
+        typeof extension === "function") {
+        if (!existingMethodImpl.extensions) {
+            throw new Error("Trying to extend overridden method,"+
+                " please use util.js:hookTo() to extend it");
+        }
+        if (existingMethodImpl.extensions.indexOf(extension) >= 0) {
+            throw new Error("Trying to register same extension method twice");
+        }
+        if (ignoreSameName) {
+            for (var i=0; i < existingMethodImpl.extensions.length; i++) {
+                if (existingMethodImpl.extensions[i].constructor.name ===
+                    extension.constructor.name) {
+                    existingMethodImpl.extensions.splice(i,1);
+                }
+            }
+        }
+        existingMethodImpl.extensions.push(extension);
+        prototypeObject[methodName] = generateFunctionFromExtensions(
+            existingMethodImpl.extensions);
+    } else if (typeof existingMethodImpl === "undefined") {
+        if (typeof extension === "function") {
+            prototypeObject[methodName] = generateFunctionFromExtensions(
+                [extension]);
+        } else {
+            prototypeObject[methodName] = generateFunctionFromExtensions([]);
+        }
+    } else {
+        throw new Error("Trying to wrap " + typeof existingMethodImpl +
+            ", expecting a function or undefined");
+    }
+}
+
+function unhookFrom(prototypeObject, methodName, extension) {
+    if (!prototypeObject[methodName].extensions) {
+        throw new Error("Trying to unhook from method, that was not extended");
+    }
+    var index = prototypeObject[methodName].extensions.indexOf(extension);
+    if (index < 0) {
+        throw new Error("Trying to unhook extension function " + extension +
+            " that was not hooked to current prototype, extensions that were" +
+            " hooked are: " + prototypeObject[methodName].extensions);
+    }
+    prototypeObject[methodName].extensions.splice(index,1);
+    prototypeObject[methodName] = generateFunctionFromExtensions(
+        prototypeObject[methodName].extensions);
+}
+
+function generateFunctionFromExtensions(extensions) {
+    var currentImpl = null;
+    if (extensions.length === 0) {
+        currentImpl = function () {};
+    } else if (extensions.length === 1) {
+        var singleImpl = extensions[0];
+        currentImpl = function () {
+            singleImpl.apply(this, arguments);
+        };
+    } else {
+        currentImpl = extensions[0];
+        for (var i = 1; i < extensions.length ; i++) {
+            var previousImpl = currentImpl;
+            currentImpl = (function () {
+                var prev = previousImpl;
+                var next = extensions[i];
+                return function () {
+                    prev.apply(this, arguments);
+                    next.apply(this, arguments);
+                };
+            }());
+        }
+    }
+    currentImpl.extensions = extensions;
```

Comment:
Is there a way to do this without adding new properties to functions? (Which changes their hidden class and opts out of function optimizations for .call and .apply for instance)